### ISSUE TITLE
Debugging session

### ISSUE THREAD

**Issue by mikeal** [mikeal/nodeconf2014#11]
@othiym23 I know we talked about this in person but I wanted to try to get in writing some of the ideas we already discussed as well as people you'd like to bring in to help.

**Comment by othiym23** [comment]
Here's where I think we left things:

In my opinion, debugging is more of a process than a particular set of tools or libraries. As such, we _can_ fit it into the workshopper format, but we'll have to be creative about it. I'm in favor of a [code kata](http://en.wikipedia.org/wiki/Kata_%28programming%29)-based approach, where we give people a set of techniques and then a set of challenges in escalating difficulty to work through.

If we want a leaderboard or some other thing that allows people to judge their progress once they leave NodeConf, we could borrow some tricks from the security community and set it up as a capture the flag-type experience. This is a little tricky if we're just doing JavaScript, but I'm pretty sure only 2 people actually made it through my final streams exercise from last year, so I'm confident I can come up with some pretty nightmarish scenarios for higher difficulty levels.

There's a lot of ground to cover, so it makes sense to choose what kind of debugging is the most important to present to a NodeConf audience. There's four basic kinds of debugging I can think of off the top of my head. They may overlap somewhat:
1. Why is my program doing this / what is my program doing?
2. Why is my program crashing?
3. Why is my program using so much memory?
4. Why is my program so slow?

I'm much stronger on 1, 2, and 4 than I am on 3, but I'm actually pretty agnostic about which are most appropriate for NodeConf. I want to focus it narrowly enough that I can slow down and not spend the entire session speaking high-velocity nerd, which I think caused problems for a lot of people at some of the more technical sessions last year (including mine -- see the above about the hardest example).

Here's the tools I do think it would be useful to mention (in rough order):
- `console.log` and `console.dir`
- `node debug` / `mocha debug`
- `node-inspector`
- refactoring and testing as debugging activities
- `longjohn`, `trycatch`, and other long stacktrace modules
- `heapdump` and using Chrome Developer Tools with Node
- basic DTrace (i.e. the pieces of DTrace that will work on all of SmartOS, OS X, and Linux)
- strategic monkeypatching for observability
- profiling / `node-webkit-agent`
- `toobusy`

Some of this may be subsumed by the `tracing` module @tjfontaine is adding to Node 0.12 -- not completely clear how much is going to make it in there.

And here's some stuff that would be fun to tack onto the end for advanced learners (like me (I hope)!):
- observing the internal state of V8 using mdb or gdb
- writing your own tracers or long-stacktrace modules with `asyncListener`
- flame graphs as a diagnostic tool
- tips'n'tricks for debugging unexpected asynchronous behavior, which could easily be its own session. It's definitely taken years off my life.

Here's my laundry list of people I think would be good to have on hand for this, in no particular order:
- anybody on core (feat. Walmart Leak Debugging Hero, TJ Fontaine)
- Jacob Groundwater
- Nick Niemeir
- Tracy Abrahms
- the usual crowd from Joyent
- Jo Liss (tentative -- right now she's not super interested in transatlantic travel in July, but will let me know if her plans change)
- Ben Acker, Wyatt Preul, and / or Eran Hammer
- Domenic Denicola, of course

**Comment by mikeal** [comment]
I don't know if you saw the format thread but this is the basics:
- Day 1 you'll only have 30 minutes. They should be able to make it through the first 2-3 challenges in the workshopper in that time. If they don't leave feeling like they have a lot more to do then the session did not succeed :)
- Day 2 people will be able to choose what sessions they want to return to in order to knock out more of the challenges where mentors are available.

Almost everyone will leave the conference with many more workshopper challenges to complete later.

Since the first 30 minute session is trying to "hook" them I wonder if we shouldn't jump straight to something like node-inspector. Linearly it would make sense to start with simpler things like console.log but given the time constraints you may want to just skip it or integrate it in to the node-inspector stuff.

**Comment by othiym23** [comment]
I totes saw the format thread, and yeah, my assumption was that there would be more information and problems in the materials than most people would be able to get through the whole time they're at NodeConf, unless they were extraordinarily motivated. My point about focus was that it's going to be tough to get _any_ of this jammed usefully into 30 minutes, and so keeping the scope as narrow as possible is pretty important.

> Linearly it would make sense to start with simpler things like console.log but given the time constraints you may want to just skip it or integrate it in to the node-inspector stuff.

The problem is that `node-inspector` is actually a pretty bad tool for serious debugging work because of how often you end up starting and stopping your tests. It's also not really available on production environments. My suggestion would be a combination of `tape` and `node debug`, if that's the route you want to go down (running the debugger on reduced test cases may not work for everybody, but it's done wonders for me).

**Comment by othiym23** [comment]
So, what are the things that other counselors think would be good to cover with debugging? What are the challenges you face most frequently? What do you wish you'd known how to do sooner? What tools do you find most useful?

@nvcexploder @chrisdickinson @thlorenz @Raynos I'm looking in particular at you.

**Comment by Raynos** [comment]
Here are some issues where I would want to know debugging techniques:
- I took a heapdump, now how do I make sense of it ?
- My node process is busy. How do I find out whether my event loop is too busy ?
- My node process has too many open fds, how do I find out where my fd leak is ?

Basically how to debug real production issues.

**Comment by othiym23** [comment]
I have the framework together and a huge heap of text at [`bug-clinic`](https://github.com/othiym23/bug-clinic). I'm furiously writing exercises and fixing bugs, but any help that anybody wants to give me, in whatever form, is super welcome.

**Comment by Raynos** [comment]
### TRIAGE

triage mentions and recommends `console.dir`, however i believe `console.log` also calls `util.inspect`

```
$ node
> console.log({ foo: 'bar' })
{ foo: 'bar' }
undefined
> console.dir({ foo: 'bar' })
{ foo: 'bar' }
```

> In day-to-day debugging, console.error, console.dir and console.trace
> are the most useful of these functions.

Since `console.dir` and `console.log` and `console.error` all print objects, is it worth mentioning `console.dir`, it feels redundant

**Comment by Raynos** [comment]
### DISINFECT

> Put it in a file named jshintrc, and then run jshint with `jshint --config jshintrc`

If this file is called `.jshintrc` then running `jshint` in that folder will pick it up automatically

**Comment by Raynos** [comment]
### TRACES

As mentioned on twitter, I couldn't get `longjohn` to actually print a proper long stacktrace, and only `stackup` worked. Might be best to recommend `stackup` instead of `longjohn`

**Comment by Raynos** [comment]
### TEST

> You should be able to run your test cases directly from the command line, without the use of a special-purpose binary.

Mentioning `node debug test/foo.js` or `mocha --debug test/foo.js` is really awesome. Mentioning that you can do `debugger;` to add breakpoints is really cool way to pause your tests and see whats going on.

**Comment by Raynos** [comment]
### INTERVENE

> and you should be careful to scrub out your debugger statements when you're done!)

jshint warns about debugger statements if configured properly. Might be worth a mention

**Comment by Raynos** [comment]
### BIOSPY

> Interpreting heap snpshots (especially when diffing them) is a bit of a black art, and there's no substitute for practice.

I myself have not learned this black art yet. Any tips you can give on how to intepret heaps into the v8 dev tools are much appreciated.

**Comment by Raynos** [comment]
### POSTMORTEM

> TODO: stick trevnorris instructions for using gdb-jit here

Learning how to use gdb on a core dump is something i'm very interested in :)

Any tips on how to get gdb or some debugger working on the latest 0.10 would be really helpful

> mdb, has recently been taught how to read core dumps from Linux. 

Linking to instructions on how to get `mdb` working in a VM somehow without paying joyent money would be super cool.

**Comment by Raynos** [comment]
### Topics missing

Here's a list of topics related to debugging that are missing. Some of these might not make sense to add.
- how to use the v8 CPU profiler
- debugging hanging processes with the event loop ( https://gist.github.com/Raynos/ef3537cfc88cd5d13319 )
- using system level debugging tools like `lsof -p PID` and `strace -p PID` to look into network traffic and system calls being made. Very useful for debugging fd leaks and such
- tips on what's save to do against productions processes
- monitoring monitoring monitoring. Sending information to statsd and have "health of my system" dashboards in graphite is god send. Feel free to replace these technologies with some better things

**Comment by othiym23** [comment]
> If this file is called .jshintrc then running jshint in that folder will pick it up automatically

yeah, I specifically don't want them to do that in this case, because I'm asking them to use very specific settings.
